// dynaform/esm5/lib/controls/color-picker/color-picker.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
var ColorPickerComponent = /** @class */ (function () {
    function ColorPickerComponent() {
    }
    /**
     * @return {?}
     */
    ColorPickerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} colorCode
     * @return {?}
     */
    ColorPickerComponent.prototype.onColorPickerChange = /**
     * @param {?} colorCode
     * @return {?}
     */
    function (colorCode) {
        this.group.get(this.field.name).setValue(colorCode);
    };
    ColorPickerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-color-picker',
                    template: "<div class=\"form-group\" [formGroup]=\"group\">\n    <label>{{field.label}}</label>\n    <input class=\"form-control color-text\" [formControlName]=\"field.name\" [value]=\"field.value\" [(colorPicker)]=\"field.value\" [style.background]=\"field.value\" [cpOKButton]=\"true\" [cpCancelButton]=\"true\" [cpOutputFormat]=\"'hex'\" [cpSaveClickOutside]=\"true\"\n        (colorPickerChange)=\"onColorPickerChange($event)\">\n</div>",
                    styles: [".color-text{font-weight:bolder;color:#fff}"]
                },] },
    ];
    /** @nocollapse */
    ColorPickerComponent.ctorParameters = function () { return []; };
    return ColorPickerComponent;
}());
export { ColorPickerComponent };
if (false) {
    /** @type {?} */
    ColorPickerComponent.prototype.field;
    /** @type {?} */
    ColorPickerComponent.prototype.group;
}
