// dynaform/esm2015/lib/controls/color-picker/color-picker.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
export class ColorPickerComponent {
    constructor() { }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} colorCode
     * @return {?}
     */
    onColorPickerChange(colorCode) {
        this.group.get(this.field.name).setValue(colorCode);
    }
}
ColorPickerComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-color-picker',
                template: `<div class="form-group" [formGroup]="group">
    <label>{{field.label}}</label>
    <input class="form-control color-text" [formControlName]="field.name" [value]="field.value" [(colorPicker)]="field.value" [style.background]="field.value" [cpOKButton]="true" [cpCancelButton]="true" [cpOutputFormat]="'hex'" [cpSaveClickOutside]="true"
        (colorPickerChange)="onColorPickerChange($event)">
</div>`,
                styles: [`.color-text{font-weight:bolder;color:#fff}`]
            },] },
];
/** @nocollapse */
ColorPickerComponent.ctorParameters = () => [];
if (false) {
    /** @type {?} */
    ColorPickerComponent.prototype.field;
    /** @type {?} */
    ColorPickerComponent.prototype.group;
}
